src/main/java/org/jvnet/hudson/test/InboundAgentRule.java

```diff
@@ -50,32 +52,184 @@
  */
 public final class InboundAgentRule extends ExternalResource {
 
-    private final Map<String, Process> procs = new HashMap<>();
+    private final ConcurrentMap<String, Process> procs = new ConcurrentHashMap<>();
+
+    /**
+     * The options used to (re)start an inbound agent.
+     */
+    public static class Options {
+        // TODO Java 14+ use records
+
+        @CheckForNull private String name;
+        private boolean secret;
+        private boolean webSocket;
+        private boolean start = true;
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(@CheckForNull String name) {
+            this.name = name;
+        }
+
+        public boolean isSecret() {
+            return secret;
+        }
+
+        public void setSecret(boolean secret) {
+            this.secret = secret;
+        }
+
+        public boolean isWebSocket() {
+            return webSocket;
+        }
+
+        public void setWebSocket(boolean webSocket) {
+            this.webSocket = webSocket;
+        }
+
+        public boolean isStart() {
+            return start;
+        }
+
+        public void setStart(boolean start) {
+            this.start = start;
+        }
+
+        /**
+         * A builder of {@link Options}.
+         *
+         * <p>Instances of {@link Builder} are created by calling {@link
+         * InboundAgentRule.Options#newBuilder}.
+         */
+        public interface Builder {
```

Comment:
would you be able to explain why you went for an interface for a builder? It's not a pattern I'm familiar with (of course there's many slightly different ways to do builders).

Reply:
I also struggled to understand this (looking into https://github.com/jenkinsci/jenkins-test-harness/pull/514#issuecomment-1319716770). Conventionally builders would be concrete `final` classes. AFAICT the only implementation is in fact `BuilderImpl`?

Reply:
If you do not like it, then change it.